### PR TITLE
Automated cherry pick of #104833 (1.22): Refine locking in API Priority and Fairness config controller

### PR DESCRIPTION
**Pull request by shyamjvs**
Cherry pick of #104833 on release-1.22.

#104833: Refine locking in API Priority and Fairness config controller

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (shyamjvs) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:136]: This was wrongly added with the rebase, lemme fix it.

Comment (shyamjvs) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:136]: Sorry, ignore above comment. I was looking at the cherrypick for older versions 1.21 and 1.20.

**Review by MikeSpreitzer — APPROVED** [review]
/LGTM

**Comment by MikeSpreitzer** [comment]
@shyamjvs : thanks for doing this

**Comment by MikeSpreitzer** [comment]
Release managers note: we are proposing to cherry-pick this because it significantly reduces the lock contention in one of the request filters in the apiserver, thus making a performance improvement for all requests.

**Comment by shyamjvs** [comment]
@savitharaghunathan Given you were the release manager for 1.22, could you help approve this cherrypick?

**Comment by puerco** [comment]
Thank you for the cherry picks @shyamjvs  !
When filing cherry picks, be sure to ping the @kubernetes/release-managers group instead of the release leads. The release leads are not release managers and cannot approve the cherry-picks :)
/lgtm

**Comment by MikeSpreitzer** [comment]
Why is the bot not responding?   Let's prod it.
/hold

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:301]: is it correct that we're calling UpdateObservations() under a read lock?

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:301]: Yes.  See the writes into the observers are not what's covered by the write locking here.

**Pull request merged**
